R/convertFlowCore.R

```diff
@@ -0,0 +1,177 @@
+        c(gate$model$rectangle$x1, gate$model$rectangle$x2),
+        c(gate$model$rectangle$y1, gate$model$rectangle$y2)),
+      c(gate$xChannel, gate$yChannel))),
+    "EllipseGate" = {
+      ellipse <- gate$model$ellipse
+      x <- unlist(ellipse$center)[1]
+      y <- unlist(ellipse$center)[2]
+      points <- getEllipsePoints(ellipse$angle, ellipse$major, ellipse$minor, x, y)
+      points <- t(data.frame(points))
+
+      result <- fitEllipsePoints(points)
+
+      cov <- result$covar
+      colnames(cov) <- c(gate$xChannel, gate$yChannel)
+      rownames(cov) <- c(gate$xChannel, gate$yChannel)
+
+      return(flowCore::ellipsoidGate(filterId = gate$name, .gate = cov, mean = c(result$x, result$y)))
+    },
+    "PolygonGate" = {
+      m <- gate$model$polygon$vertices
+      if (is.null(dim(m))) {
+        m <- m[[1]]
+      }
+      colnames(m) <- c(gate$xChannel, gate$yChannel)
+      flowCore::polygonGate(filterId = gate$name, m)
+    },
+    "QuadrantGate" = stop("This gate has no representation in flowCore"),
+    "SplitGate" = stop("This gate has no representation in flowCore")
+  )
+}
+
+#' Scale a flowCore gate or flowDensity filter
+#'
+#' Applies a scale to points from a flowCore gate, so that it may
+#' be saved to CellEngine.
+#'
+#' @param gate The flowCore gate points (i.e. polygonGate@boundaries)
+#' @param scaleSet The CellEngine ScaleSet to apply.
+#' @export
+#' @examples
+#' \dontrun{
+#' sngl <- flowDensity(file, params)
+#' flowGate <- sngl@filter
+#' scaleSet <- getScaleSets("5d2f8b4b21fd0676fb3a6a8c")
+#' gatePoints <- scaleFlowCoreGate(flowGate, scaleSet)
+#' createPolygonGate(id, colnames(flowGate)[1], colnames(flowGate)[2], 'my gate', vertices=gatePoints)
+#' }
+scaleFlowCoreGate <- function (gate, scaleSet) {
+  xChannel <- colnames(gate)[1]
+  yChannel <- colnames(gate)[2]
+  scales <- data.frame(scaleSet$scales)
+  ss <- scales$scale
+  rownames(ss) <- scales$channelName
+
+  scaleX <- ss[xChannel,]
+  scaleY <- ss[yChannel,]
+
+  scaledGate = cbind(applyScale(scaleX, gate[,xChannel]), applyScale(scaleY, gate[,yChannel]))
+  colnames(scaledGate) <- c(xChannel, yChannel)
+  gate[,xChannel] <- unlist(scaledGate[,xChannel])
+  gate[,yChannel] <- unlist(scaledGate[,yChannel])
+
+  return(lapply(seq_len(nrow(gate)), function(i) gate[i,]))
+}
+
+scaleEllipseCenter <- function (center, scaleSet) {
+  xChannel <- names(center)[1]
+  yChannel <- names(center)[2]
+  scales <- data.frame(scaleSet$scales)
+  s <- scales$scale
+  rownames(s) <- scales$channelName
+
+  scaleX <- s[xChannel,]
+  scaleY <- s[yChannel,]
+
+  scaledCenter = cbind(applyScale(scaleX, center[xChannel]), applyScale(scaleY, center[yChannel]))
+  names(scaledCenter) <- c(xChannel, yChannel)
+  center[xChannel] <- unlist(scaledCenter[xChannel])
+  center[yChannel] <- unlist(scaledCenter[yChannel])
+
+  return(center)
+}
+
+#' Convert flowCore to gate
+#'
+#' Converts points in a flowCore gate or flowDensity filter to a CellEngine
```

Comment:
Slightly hesitant to include flowDensity here. There are tons of algorithms out there and this sort of begs us to handle them all if we handle any of them. Maybe just put it in a cookbook recipe for now?

Reply:
AFAIK, a flowDensity filter is always equivalent to a flowCore gate, so it was easy enough to include here. That said, this won't apply necessarily well to other data structures. I could pull the flowCore-specific parts out of this, and only operate on the underlying data, but....